networkx/algorithms/cluster.py

```diff
@@ -42,22 +44,49 @@ def triangles(G, nodes=None):
     6
     >>> print(nx.triangles(G))
     {0: 6, 1: 6, 2: 6, 3: 6, 4: 6}
-    >>> print(list(nx.triangles(G, (0, 1)).values()))
+    >>> print(list(nx.triangles(G, [0, 1]).values()))
     [6, 6]
 
     Notes
     -----
-    When computing triangles for the entire graph each triangle is counted
-    three times, once at each node.  Self loops are ignored.
+    Self loops are ignored.
 
     """
-    # If `nodes` represents a single node in the graph, return only its number
-    # of triangles.
-    if nodes in G:
-        return next(_triangles_and_degree_iter(G, nodes))[2] // 2
-    # Otherwise, `nodes` represents an iterable of nodes, so return a
-    # dictionary mapping node to number of triangles.
-    return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+    if nodes is not None:
+        # this checks if `nodes` could potentially refer to a single node
+        # If `nodes` represents a single node, return only its number of triangles
+        # (if it's in the graph) or zero (if the node is not in the graph).
+        if callable(getattr(nodes, "__hash__")):
```

Comment:
Is there a way to avoid diving into the object model? That might help with readability. For example:

```suggestion
        # If `nodes` represents a single node, return only its number of triangles
        if nodes in G:
```

Note also that `triangles` currently raises when `nodes` represents a single node that is not in the Graph, so I think this might've changed behavior for that corner case. A test should be added just to be sure!

Reply:
Interesting, so suppose `nodes` is a list, then `if nodes in G` will be False, but we still don't know if it's because `nodes` is a container or because the single node in `nodes` is not in the graph... that's why it seems better to first check whether `nodes` is a container vs single node.

re: raising for a single node not in the graph, do you mean this (raising a KeyError or NetworkxError) is the desired behavior (to test against)?

Reply:
> Interesting, so suppose nodes is a list, then if nodes in G will be False, but we still don't know if it's because nodes is a container or because the single node in nodes is not in the graph... that's why it seems better to first check whether nodes is a container vs single node.

Unfortunately I think this is a fundamental ambiguity, since it's perfectly valid to have nodes that are iterable that meet this criteria.

In fact, the docstring itself is ambiguous: the parameter description says that `nodes` is supposed to be a container of nodes, but the examples section shows an example of a single node :shrug: .

Either way this is a clear corner case. For example, consider the following:

```python
>>> G = nx.Graph()
>>> G.add_node((0, 0))  # Nodes are 2-tuples, perhaps representing 2D coordinates for example
>>> G.add_node((2, 3))
# Let's assume the user intends (0, 1) to represent a single node, that happens not to be in G.
# What happens in this case?
>>> nx.triangles(G, (0, 1))  
```
On main:
```python
>>> nx.triangles(G, (0, 1))
{}
```
This PR:
```python
>>> nx.triangles(G, (0, 1))
Traceback (most recent call last)
   ...
StopIteration
```

Now as to which behavior is "correct" - I'd argue neither! But either way this is also what I was referring to with this being an untested corner case.

Reply:
Oops, I fixed the `StopIteration` bug. Thank you for pointing this out! I also added a test, for now assuming that the correct return is 0 triangles for a node not in the graph.

Reply:
Also, re: ambiguity: isn't this resolved if we check whether `nodes` could be a key in a python dict (so checking if it's hashable)? This way even if the object is iterable, we know whether it could potentially be a key in the dict (and hence a node in the graph, ignoring special case of `None`)

Reply:
> isn't this resolved if we check whether nodes could be a key in a python dict (so checking if it's hashable)?

The ambiguity arises from the fact that `nodes` can be interpreted as either a single node or a collection of nodes. When the user specifies `nodes=(0, 1)`, this can either be interpreted as: a collection containing nodes `0` and `1`; or a single node, the 2-tuple `(0, 1)`.

This ambiguity *wouldn't* exist if there were stronger enforcement of the parameter description (i.e. `nodes : container of nodes, optional (default = all nodes in G)`), but unfortunately [that ship has sailed](https://github.com/networkx/networkx/blob/0f91550007fd3a95261d858b1a6a623ef8bda38a/networkx/algorithms/cluster.py#L41).

The question then becomes what to do about it. Should we drop support for treating `nodes` as a single node? That would resolve the ambiguity but would change behavior and likely affect users. Should we add a bunch of input validation boilerplate?

Generally I think this is a larger discussion/separate issue - I only raised it here because I noticed that the proposed changes would change behavior for this (and potentially other) untested corner case(s).

Reply:
If possible, can we leave the behavior the same for the following cases:
- a single node in the graph (if nodes in G: return a single integer value)
- a container that happens to hold at least one node of the graph (and return a dict keyed by node to the number)
- None as the default and leads to a dict keyed by each node of G to the number for that node.

It'd be great if the ambiguous corner cases could either be treated better or the same as they are currently treated.